cpp/mcap/include/mcap/mcap.inl

```diff
@@ -0,0 +1,168 @@
+  write(message.publishTime);
+  write(message.recordTime);
+  write(message.data, message.dataSize);
+
+  return std::nullopt;
+}
+
+std::optional<std::error_code> McapWriter::write(const mcap::Attachment& attachment) {
+  if (!stream_) {
+    return std::make_optional(make_error_code(ErrorCode::NotOpen));
+  }
+
+  const uint64_t recordSize =
+    4 + attachment.name.size() + 8 + 4 + attachment.contentType.size() + attachment.dataSize;
+
+  write(OpCode::Attachment);
+  write(recordSize);
+  write(attachment.name);
+  write(attachment.recordTime);
+  write(attachment.contentType);
+  write(attachment.data, attachment.dataSize);
+
+  return std::nullopt;
+}
+
+// Private methods /////////////////////////////////////////////////////////////
+
+namespace internal {
+
+uint32_t KeyValueMapSize(const KeyValueMap& map) {
+  uint32_t size = 0;
+  for (const auto& [key, value] : map) {
+    size += 4 + key.size() + 4 + value.size();
+  }
+  return size;
+}
+
+}  // namespace internal
+
+void McapWriter::writeMagic() {
+  stream_->write(Magic, sizeof(Magic));
+}
+
+void McapWriter::write(const mcap::Header& header) {
+  const uint32_t metadataSize = internal::KeyValueMapSize(header.metadata);
+  const uint64_t recordSize =
+    4 + header.profile.size() + 4 + header.library.size() + 4 + metadataSize;
+
+  write(OpCode::Header);
+  write(recordSize);
+  write(header.profile);
+  write(header.library);
+  write(header.metadata, metadataSize);
+}
+
+void McapWriter::write(const mcap::Footer& footer) {
+  write(OpCode::Footer);
+  write(uint64_t(12));
+  write(footer.indexOffset);
+  write(footer.indexCrc);
+}
+
+void McapWriter::write(const mcap::ChannelInfo& info) {
+  const uint32_t userDataSize = internal::KeyValueMapSize(info.userData);
+  const uint64_t recordSize = 2 + 4 + info.topicName.size() + 4 + info.encoding.size() + 4 +
+                              info.schemaName.size() + 4 + info.schema.size() + 4 + userDataSize +
+                              4;
+  const uint32_t crc = 0;
+
+  write(OpCode::ChannelInfo);
+  write(recordSize);
+  write(info.channelId);
+  write(info.topicName);
+  write(info.encoding);
+  write(info.schemaName);
+  write(info.schema);
+  write(info.userData, userDataSize);
+  write(crc);
+}
+
+void McapWriter::write(const std::string_view str) {
+  write(uint32_t(str.size()));
+  stream_->write(str.data(), str.size());
+}
+
+void McapWriter::write(OpCode value) {
+  stream_->write(reinterpret_cast<const char*>(&value), sizeof(value));
+}
+
+void McapWriter::write(uint16_t value) {
+  stream_->write(reinterpret_cast<const char*>(&value), sizeof(value));
+}
+
+void McapWriter::write(uint32_t value) {
+  stream_->write(reinterpret_cast<const char*>(&value), sizeof(value));
+}
+
+void McapWriter::write(uint64_t value) {
+  stream_->write(reinterpret_cast<const char*>(&value), sizeof(value));
+}
+
+void McapWriter::write(uint8_t* data, uint64_t size) {
+  stream_->write(reinterpret_cast<const char*>(data), size);
+}
+
```

Comment:
Need endianness checks, or (my preference) just use bitwise shift and `&` to write out the bytes you want to write in the order you want them

Reply:
Can we just static_assert() on little-endian?

Reply:
👍
I do wonder what would go wrong in our code if CHAR_BIT≠8? I guess all these sizeof()s would be wrong?